File: Sort/1235. Maximum Profit in Job Scheduling/1235. Maximum Profit in Job Scheduling.java
class Solution {
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        // sorted data structure for endtime
        PriorityQueue<int[]> pq = new PriorityQueue<>( (a, b) -> (a[0] -b[0]) );
        int N = startTime.length;
        int[] sortedEndTime = new int[N+1];
        for (int i = 0; i < N; i++) {
            pq.offer(new int[] {endTime[i], startTime[i], profit[i]});
            sortedEndTime[i+1] = endTime[i];
        }
        Arrays.sort(sortedEndTime);
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 0);
        // for the current earliest endtime to handle
        int curVal = 0;
        while ( !pq.isEmpty() ) {
            int[] info = pq.poll();
            int endT = info[0];
            int startT = info[1];
            int val = info[2];
            
            int beforeIdx = Arrays.binarySearch(sortedEndTime, startT);
            int prevEnd = beforeIdx < 0 ? sortedEndTime[(-1)*beforeIdx-2] : sortedEndTime[beforeIdx];
            curVal = Math.max(curVal, map.get(prevEnd) + val);
            map.put(endT, curVal);
        }
        return curVal;
    }
}

// -----------
//     ----------
//       -------------
//    -----------e`
// ------e 
//           s--------t
// dp[t] : max (dp[nearestEndTimeForCurrentMeeting], dp[lastEndTimeBeforeStartTime] + val[t]); // not select and select

File: Sort/1235. Maximum Profit in Job Scheduling/1235. Maximum Profit in Job Scheduling.cpp
class Solution {
public:
    int jobScheduling(vector<int>& startTime, vector<int>& endTime, vector<int>& profit) {
        vector<vector<int>>jobs;
        for (int i = 0; i < startTime.size(); i++) {
            jobs.push_back( {endTime[i], startTime[i], profit[i]} );
            jobs.push_back( {startTime[i], 0, INT_MIN} );
        }
        sort(jobs.begin(), jobs.end()); // sort two times
        
        unordered_map<int, int> time2val;
        time2val[0] = 0;
        
        int curVal = 0;
        for (auto job : jobs) {
            int end = job[0];
            int start = job[1];
            int val = job[2];
            
            time2val[end] = max(curVal, time2val[start] + val);
            curVal = time2val[end];
        }
        return curVal;
    }
};
